### PR TITLE
ante: check that sender is a EOA  #73

### LINKED ISSUE #73: ante: check that sender is a EOA
Actionable Items:

- Add check on `AnteHandler` to check if the user is an EOA
  - Note: need to pass an infinite gas meter to call `evmKeeper.GetCodeSize` and then reset to the original gas meter

Ref: https://github.com/ethereum/go-ethereum/pull/23002

### PR DESCRIPTION
**Pull request by bwhour**
closes  #73

### CODE REVIEW

File: app/ante/eth.go [hunk 1]
```diff
@@ -122,6 +123,13 @@ func (avd EthAccountVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx
 			return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "from address cannot be empty")
 		}
 
+		// check whether the sender address is EOA
+		fromAddr := common.BytesToAddress(from)
+		if codesize := avd.evmKeeper.GetCodeSize(fromAddr); codesize != 0 {
```
Comment (fedekunze) [inline app/ante/eth.go:128]: actually, a better approach would be to query for the block hash and return an error if it's equal to the `EmptyCodeHash`

Comment (bwhour) [reply app/ante/eth.go:128]: thanks for your comment, i saw the comment from @albertchon . it will be more efficient to just check the code hash. 
i just think  using GetCodeSize would have better code encapsulation and readability.
I will change it to GetCodeHash.

Comment (fedekunze) [reply app/ante/eth.go:128]: yeah, it's just that internally `GetCodeSize` queries for `GetCode` which itself queries `GetCodeHash`. So we can avoid a few db reads from just using the proposed method 👍

File: app/ante/eth.go [hunk 2]
```diff
@@ -147,6 +148,14 @@ func (avd EthAccountVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx
 			)
 		}
 
+		// check whether the sender address is EOA
+		fromAddr := common.BytesToAddress(from)
+		codeHash := avd.evmKeeper.GetCodeHash(fromAddr)
+		if codeHash == common.BytesToHash(evmtypes.EmptyCodeHash) {
+			return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress,
+				"the sender is not EOA: address <%v>, codeHash is empty", fromAddr)
```
Comment (fedekunze) [inline app/ante/eth.go:156]:
```suggestion
				"the sender %s is not EOA, codeHash is empty", fromAddr)
```

File: x/evm/keeper/statedb.go [hunk 3]
```diff
@@ -269,9 +270,17 @@ func (k *Keeper) SetCode(addr common.Address, code []byte) {
...
 func (k *Keeper) GetCodeSize(addr common.Address) int {
-	return len(k.GetCode(addr))
+	// TODO (@khoslaventures): suppose this call to GetCode fails. How does
+	// GetCodeSize know without an error return - is nil is always an error case?
+	// Can we be more explicit?
+	code := k.GetCode(addr)
+	if code != nil {
+		return len(code)
+	}
+	return 0
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:283]: `len(nil)` is always 0, so you can undo this change

File: x/evm/keeper/statedb.go [hunk 4]
```diff
@@ -216,6 +216,7 @@ func (k *Keeper) GetCode(addr common.Address) []byte {
 	code := store.Get(hash.Bytes())
 
 	if len(code) == 0 {
+		// TODO (@khoslaventures): why not error?
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:219]: because the `StateDB` interface doesn't error

File: app/ante/eth.go (hunk 2 above)
Comment (fedekunze) [reply app/ante/eth.go:156]: also you can `stacktrace.Propagate` the error

File: x/evm/keeper/statedb.go (hunk 4 above)
Comment (fedekunze) [reply x/evm/keeper/statedb.go:219]: maybe we should start a discussion on Geth...

File: x/evm/keeper/statedb.go [hunk 5]
```diff
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"math/big"
 
+	evmtypes "github.com/cosmos/ethermint/x/evm/types"
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:8]:
```suggestion
```

File: x/evm/keeper/statedb.go [hunk 6]
```diff
@@ -216,6 +217,7 @@ func (k *Keeper) GetCode(addr common.Address) []byte {
 	code := store.Get(hash.Bytes())
 
 	if len(code) == 0 {
+		// TODO (@khoslaventures): why not error?
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:220]:
```suggestion
```

File: x/evm/keeper/statedb.go [hunk 7]
```diff
@@ -228,6 +230,9 @@ func (k *Keeper) GetCode(addr common.Address) []byte {
 
 // SetCode calls CommitStateDB.SetCode using the passed in context
 func (k *Keeper) SetCode(addr common.Address, code []byte) {
+	if bytes.Equal(code, evmtypes.EmptyCodeHash) {
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:233]:
```suggestion
	if bytes.Equal(code, types.EmptyCodeHash) {
```
